# Content

* 1. Import Libraries

* 2. Creating user Defined Functions

* 3. Set working Path

* 4. Import Data
    * 4.1 Create Subset

* 5. IF-statement
    * 5.1 If-else statement
    * 5.2 If-statement with .loc() function
    
* 6. Apply on the entire dataframe
    * 6.1 F-Loop function
    
* 7. Exercise 4.7
    * 7.1 Task 2
    * 7.2 Task 3: Check the values of new column for accuracy
    * 7.3 Task 4: Find busiest day of the week
    * 7.4 Task 5: Print Frequency of column 'busiest_period_pf_day'
    
* 8 Export


In [1]:
#import the libraries
import pandas as pd
import numpy as np
import os

# How to create user defined functions

In [2]:
# Define a function
def add_num(x,y):
    sum=x+y
    return sum

In [3]:
#run the function above
add_num(9,11)

20

# 3. Set work Path

In [2]:
# set path for work
path= r'/Users/phongsathorn/Documents/Data Analytics/CareerFoundry/Achievement 4 - Python/InstaCart/Instacart Business Analysis'

# 4. Import Data

In [3]:
# import ords_prods_merge file
df=pd.read_pickle(os.path.join(path,'02 Data','Prepared Data','ords_prods_merge.pkl'))

## 4.1 Create Subset

In [5]:
# Create a subset of the df, because the dataframe is too big
# the first one million rows in the dataframe.
df=df[:1000000]

In [6]:
# see the dataframe .head()
df.head()

order_id  user_id  order_number  orders_day_of_week  time_of_day  \
0   2539329        1             1                   2            8   
1   2398795        1             2                   3            7   
2    473747        1             3                   3           12   
3   2254736        1             4                   4            7   
4    431534        1             5                   4           15   

   days_since_previous_order  product_id  add_to_cart_order  reordered  \
0                        NaN         196                  1          0   
1                       15.0         196                  1          1   
2                       21.0         196                  1          1   
3                       29.0         196                  1          1   
4                       28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge  
0         Soda        77              7     9.0   both  
1         Soda        77              7     9.0   both  
2         Soda        77              7     9.0   both  
3         Soda        77              7     9.0   both  
4         Soda        77              7     9.0   both

The goal is to create function with the following condition:

1.
If the item’s price is lower than or equal to $5, it will be labeled a “low-range product.”

2.
If the item’s price is above $5 but lower than or equal to $15, it will be labeled a “mid-range product.”

3.
If the item’s price is above $15, it will be labeled a “high-range product.”

# 5. If-statements

## 5.1 IF-Else statement

In [8]:
# create the IF statement that fulfills all 3 conditions
# Define a flage 'price_label'

def price_label(row):

  if row['prices'] <= 5:
    return 'Low-range product'
  elif (row['prices'] > 5) and (row['prices'] <= 15):
    return 'Mid-range product'
  elif row['prices'] > 15:
    return 'High range'
  else: return 'Not enough data'

In [9]:
def price_label(row):
# Attempt to convert the price to a float, if it fails, it means the data is not a valid number
price = float(row['prices'])
except (ValueError, TypeError):
return 'Not enough data'

# After successfully converting to float, we proceed with comparison
if price <= 5:
return 'Low-range product'
elif price <= 15:
return 'Mid-range product'
else: # This covers any price above 15
return 'High-range product'

IndentationError: expected an indented block after function definition on line 1 (2841084887.py, line 3)

In [9]:
# Use the function
# Creating a new column of 'price_range' in df, that will contain all the price labels from the previous
#function using the IF statement

df['price_range'] = df.apply(price_label, axis=1)

In [10]:
# count values in the new column of 'price_range'
df['price_range'].value_counts()

price_range
Mid-range product    756450
Low-range product    243550
Name: count, dtype: int64

**The missing "High-range products" means that, within this subset of 1 mio products, no products are higher than 15$.**

In [11]:
# Finding out the max. price in the subset
df['prices'].max()

14.8

## 5.2 If statement with loc() function

**Using the loc() function, it is also possible to get the same results as IF-statments**. 

You can translate the loc() function into this:

If  =   df.loc[df['column name'] < 15,


Then =  'new column name that contains the results'] = 'High-range product'

In [12]:
df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [13]:
df.loc[(df['prices'] <= 15) & (df['prices'] > 5), 'price_range_loc'] =     'Mid-range product' 

In [14]:
df.loc[df['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [15]:
# Count the values of the loc() function
df['price_range_loc'].value_counts()

price_range_loc
Mid-range product    756450
Low-range product    243550
Name: count, dtype: int64

# 6. Repeat the same step with the entire dataframe and not only the subset

In [16]:
# import ords_prods_merge file
ords_prods_merge=pd.read_pickle(os.path.join(path,'02 Data','Prepared Data','ords_prods_merge.pkl'))

In [17]:
ords_prods_merge.loc[ords_prods_merge['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [18]:
ords_prods_merge.loc[(ords_prods_merge['prices'] <= 15) & (ords_prods_merge['prices'] > 5), 'price_range_loc'] ='Mid-range product' 

In [19]:
ords_prods_merge.loc[ords_prods_merge['prices']<=5, 'price_range_loc']= 'Low-range products'

In [20]:
ords_prods_merge['price_range_loc'].value_counts()

price_range_loc
Mid-range product     21860860
Low-range products    10126321
High-range product      417678
Name: count, dtype: int64

# 6.1 F-LOOP function

Using the instaChart dataframe, You’ll create a new column in your ords_prods_merge dataframe that summarizes how busy each day of the week is.

In [21]:
#First steps:Know on which day most orders take place
ords_prods_merge['orders_day_of_week'].value_counts()

orders_day_of_week
0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: count, dtype: int64

It seems like that the day with the most orders is #0 which is a Saturday. The least orders is during day#4, which is a Wednesday.

We can create a new column now in the dataframe and assign the respective days with 'busiest day', 'least_busiest_day' or 'regular_busy'.

Using loop with IF structure!!

The loop will run through every row in the “orders_day_of_week” column, compare its value with what you know are the busiest and slowest days, and assign it the corresponding string value.

In [22]:
# Creating the code for for-loop

result = []

for value in ords_prods_merge["orders_day_of_week"]:
  if value == 0:
    result.append("Busiest day")
  elif value == 4:
    result.append("Least busy")
  else:
    result.append("Regularly busy")

In [23]:
# Now use the list of 'result' as a new column in the dataframe
ords_prods_merge['busiest_day']=result

In [24]:
# print the frequency of this new column and cross-check it with the frequency of “orders_day_of_week”


In [25]:
ords_prods_merge['busiest_day'].value_counts(dropna=False)

busiest_day
Regularly busy    22416875
Busiest day        6204182
Least busy         3783802
Name: count, dtype: int64

**Result:** The output shows how many sales occur on each different type of day

# 7 Exercise 4.7

## 7.1 Task 2

Create a new collumn that contains 'Busies days' for the 2 most busiest days, and 'Slowest days' for the 2 least busiest day

In [26]:
# Check for .head()
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  time_of_day  \
0   2539329        1             1                   2            8   
1   2398795        1             2                   3            7   
2    473747        1             3                   3           12   
3   2254736        1             4                   4            7   
4    431534        1             5                   4           15   

   days_since_previous_order  product_id  add_to_cart_order  reordered  \
0                        NaN         196                  1          0   
1                       15.0         196                  1          1   
2                       21.0         196                  1          1   
3                       29.0         196                  1          1   
4                       28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge    price_range_loc  \
0         Soda        77              7     9.0   both  Mid-range product   
1         Soda        77              7     9.0   both  Mid-range product   
2         Soda        77              7     9.0   both  Mid-range product   
3         Soda        77              7     9.0   both  Mid-range product   
4         Soda        77              7     9.0   both  Mid-range product   

      busiest_day  
0  Regularly busy  
1  Regularly busy  
2  Regularly busy  
3      Least busy  
4      Least busy

In [27]:
# Creating the code for for-loop

result = []


for value in ords_prods_merge["orders_day_of_week"]:
    if value in (0, 1):
        result.append("Busiest days")
    elif value in (3, 4):
        result.append("Slowest days")
    else:
        result.append("Regularly day")

In [28]:
# Creating the new column for the dataframe
ords_prods_merge['Busiest Days']=result

In [29]:
# Check the dataframe
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  time_of_day  \
0   2539329        1             1                   2            8   
1   2398795        1             2                   3            7   
2    473747        1             3                   3           12   
3   2254736        1             4                   4            7   
4    431534        1             5                   4           15   

   days_since_previous_order  product_id  add_to_cart_order  reordered  \
0                        NaN         196                  1          0   
1                       15.0         196                  1          1   
2                       21.0         196                  1          1   
3                       29.0         196                  1          1   
4                       28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge    price_range_loc  \
0         Soda        77              7     9.0   both  Mid-range product   
1         Soda        77              7     9.0   both  Mid-range product   
2         Soda        77              7     9.0   both  Mid-range product   
3         Soda        77              7     9.0   both  Mid-range product   
4         Soda        77              7     9.0   both  Mid-range product   

      busiest_day   Busiest Days  
0  Regularly busy  Regularly day  
1  Regularly busy   Slowest days  
2  Regularly busy   Slowest days  
3      Least busy   Slowest days  
4      Least busy   Slowest days

## 7.2
### Task 3 Check the values of this new column for accuracy

In [30]:
ords_prods_merge['Busiest Days'].value_counts(dropna = False)

Busiest Days
Regularly day    12916111
Busiest days     11864412
Slowest days      7624336
Name: count, dtype: int64

To make sure the numbers are correct:

The sum of day 0 and 1 = 11864412, which is equal to the numbers of values in "Busiest days"

The sum of day 3 and 4 = 7624336

## 7.3
### Task 4.  Identify the busiest hours of the day

Show periods of time labeled “Most orders,” “Average orders,” and “Fewest orders" and put this in a new column

In [31]:
# count the orders during the day

ords_prods_merge['time_of_day'].value_counts()

time_of_day
10    2761760
11    2736140
14    2689136
15    2662144
13    2660954
12    2618532
16    2535202
9     2454203
17    2087654
8     1718118
18    1636502
19    1258305
20     976156
7      891054
21     795637
22     634225
23     402316
6      290493
0      218769
1      115700
5       87961
2       69375
4       53242
3       51281
Name: count, dtype: int64

Most orders are at 10, least orders are at 3, the remaining time is average orders. If devided by 3 different categories we have 8 different times of the day for each category

In [32]:
# Create for-loop if statement labeling periods of time as “Most orders,” “Average orders,” and “Fewest orders.”

result_2 = []

for value in ords_prods_merge["time_of_day"]:
  if value in [10, 11, 14, 15, 13, 12, 16, 9]:
    result_2.append("Most orders")
  elif value in [23, 6, 0, 1, 5, 2, 4, 3]:
    result_2.append("Fewest orders")
  else:
    result_2.append("Average orders")


In [33]:
# Create a new colum with the different time categories
ords_prods_merge["busiest_period_of_day"]=result_2

## 7.5
### Task 5. Print the frequency of the column'busiest_period_of_day'

In [34]:
# Check for the frequency for the new collumn
ords_prods_merge["busiest_period_of_day"].value_counts(dropna=False)

busiest_period_of_day
Most orders       21118071
Average orders     9997651
Fewest orders      1289137
Name: count, dtype: int64

# 8. Export the data as a pkl data

In [35]:
ords_prods_merge.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_derived.pkl'))